src/core/providers/vpc/qgsvirtualpointcloudprovider.cpp

```diff
@@ -0,0 +1,555 @@
+
+  return mCrs;
+}
+
+QgsRectangle QgsVirtualPointCloudProvider::extent() const
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  return mExtent;
+}
+
+QgsPointCloudAttributeCollection QgsVirtualPointCloudProvider::attributes() const
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  return mAttributes;
+}
+
+bool QgsVirtualPointCloudProvider::isValid() const
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  return !mSubLayers.isEmpty();
+}
+
+QString QgsVirtualPointCloudProvider::name() const
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  return PROVIDER_KEY;
+}
+
+QString QgsVirtualPointCloudProvider::description() const
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  return PROVIDER_DESCRIPTION;
+}
+
+QgsPointCloudIndex *QgsVirtualPointCloudProvider::index() const
+{
+  // non fatal for now -- 2d rendering of point clouds is not thread safe and calls this
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS_NON_FATAL
+
+  return nullptr;
+}
+
+qint64 QgsVirtualPointCloudProvider::pointCount() const
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  return static_cast<qint64>( mPointCount );
+}
+
+void QgsVirtualPointCloudProvider::loadIndex()
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  //no-op, there is no index
+}
+
+QVariantMap QgsVirtualPointCloudProvider::originalMetadata() const
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  return QVariantMap();
+}
+
+void QgsVirtualPointCloudProvider::generateIndex()
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  //no-op, index is always generated
+}
+
+void QgsVirtualPointCloudProvider::parseFile()
+{
+  QFile file( dataSourceUri() );
+  const QFileInfo fInfo( file );
+
+  nlohmann::json data;
+  if ( file.open( QFile::ReadOnly ) )
+  {
+    try
+    {
+      data = json::parse( file.readAll().toStdString() );
+    }
+    catch ( std::exception &e )
+    {
+      appendError( QgsErrorMessage( QStringLiteral( "JSON parsing error: %1" ).arg( QString::fromStdString( e.what() ) ), QString() ) );
+      return;
+    }
+  }
+
+  if ( data["type"] != "FeatureCollection" ||
+       !data.contains( "features" ) )
+  {
+    appendError( QgsErrorMessage( QStringLiteral( "Invalid VPC file" ), QString( "asd" ) ) );
+    return;
+  }
+
+  QSet<QString> attributeNames;
+
+  for ( const auto &f : data["features"] )
+  {
+    if ( !f.contains( "type" ) || f["type"] != "Feature" ||
+         !f.contains( "stac_version" ) ||
+         !f.contains( "assets" ) || !f["assets"].is_object() ||
+         !f.contains( "properties" ) || !f["properties"].is_object() ||
+         !f.contains( "geometry" ) || !f["geometry"].is_object() )
+    {
+      QgsDebugMsg( QStringLiteral( "Malformed STAC item: %1" ).arg( QString::fromStdString( f ) ) );
+      continue;
+    }
+
+    if ( f["stac_version"] != "1.0.0" )
+    {
+      QgsDebugMsg( QStringLiteral( "Unsupported STAC version: %1" ).arg( QString::fromStdString( f["stac_version"] ) ) );
+      continue;
+    }
+
+    QgsPointCloudSubIndex si;
+
+    nlohmann::json firstAsset = *f["assets"].begin();
```

Comment:
I think we should test first if the "assets" is empty and skip the item in such case (I know this comes from pdal_wrench, I should fix it there too :-) )

Reply:
ok, let's get the first available asset's href